code/lib/cli/src/automigrate/fixes/new-frameworks/new-frameworks.ts

```diff
@@ -0,0 +1,528 @@
+import chalk from 'chalk';
+import dedent from 'ts-dedent';
+import semver from 'semver';
+import { readConfig, writeConfig } from '@storybook/csf-tools';
+import { getStorybookInfo, loadMainConfig, rendererPackages } from '@storybook/core-common';
+
+import type { Fix } from '../../types';
+import type { PackageJsonWithDepsAndDevDeps } from '../../../js-package-manager';
+import { getStorybookVersionSpecifier } from '../../../helpers';
+import { detectRenderer } from '../../helpers/detectRenderer';
+import type { Addon } from './utils';
+import { getNextjsAddonOptions, getBuilderInfo, packagesMap } from './utils';
+import { detectFramework } from '../../helpers/detectFramework';
+
+const logger = console;
+
+interface NewFrameworkRunOptions {
+  mainConfigPath: string;
+  packageJson: PackageJsonWithDepsAndDevDeps;
+  allDependencies: Record<string, string>;
+  dependenciesToAdd: string[];
+  dependenciesToRemove: string[];
+  hasFrameworkInMainConfig: boolean;
+  frameworkPackage: string;
+  renderer: string;
+  addonsToRemove: string[];
+  frameworkOptions: Record<string, any>;
+  rendererOptions: Record<string, any>;
+  addonOptions: Record<string, any>;
+  builderConfig: string | Record<string, any>;
+  builderInfo: {
+    name: string;
+    options: Record<string, any>;
+  };
+}
+
+/**
+ * Does the user have separate framework and builders (e.g. @storybook/react + core.builder -> webpack5)?
+ *
+ * If so:
+ * - Remove the dependencies (@storybook/react + @storybook/builder-webpack5 + @storybook/manager-webpack5)
+ * - Install the correct new package e.g. (@storybook/react-webpack5)
+ * - Update the main config to use the new framework
+ * -- moving core.builder into framework.options.builder
+ * -- moving renderer options (e.g. reactOptions) into framework.options
+ * -- removing the now unnecessary fields in main.js
+ */
+export const newFrameworks: Fix<NewFrameworkRunOptions> = {
+  id: 'newFrameworks',
```

Comment:
I'm adding a comment here so this file doesn't get overlooked. Github is collapsing it by default